lib/obj.js

```diff
@@ -1,365 +1,479 @@
+    }
+    return loadMaterials(mtlPaths, objPath, options)
+        .then(function(materials) {
+            var imagePaths = getImagePaths(materials);
+            return loadImages(imagePaths, objPath, options)
+                .then(function(images) {
+                    return {
+                        nodes : nodes,
+                        materials : materials,
+                        images : images
+                    };
+                });
         });
+}
+
+function getAbsolutePath(mtlPath, objPath) {
+    if (!path.isAbsolute(mtlPath)) {
+        mtlPath = path.join(path.dirname(objPath), mtlPath);
+    }
+    return mtlPath;
+}
+
+function outsideDirectory(filePath, objPath) {
+    return (path.relative(path.dirname(objPath), filePath).indexOf('..') === 0);
+}
 
-        stream.on('end', function () {
-            resolve({
-                vertexCount: vertexCount,
-                vertexArray: vertexArray,
-                positionMin: positionMin,
-                positionMax: positionMax,
-                hasUVs: hasUVs,
-                hasNormals: hasNormals,
-                materialGroups: materialGroups,
-                materials: materials,
-                images: images
+function loadMaterials(mtlPaths, objPath, options) {
+    var materials = {};
+    return Promise.map(mtlPaths, function(mtlPath) {
+        mtlPath = getAbsolutePath(mtlPath, objPath);
+        if (options.secure && outsideDirectory(mtlPath, objPath)) {
+            options.logger('Could not read mtl file at ' + mtlPath + ' because it is outside of the obj directory and the secure flag is true. Using default material instead.');
+            return;
+        }
+        return loadMtl(mtlPath)
+            .then(function(materialsInMtl) {
+                materials = combine(materials, materialsInMtl);
+            })
+            .catch(function() {
+                options.logger('Could not read mtl file at ' + mtlPath + '. Using default material instead.');
             });
-        });
+    }).then(function() {
+        return materials;
     });
 }
 
-function getImages(inputPath, materials) {
-    // Collect all the image files from the materials
-    var images = [];
+function loadImages(imagePaths, objPath, options) {
+    var images = {};
+    return Promise.map(imagePaths, function(imagePath) {
+        if (options.secure && outsideDirectory(imagePath, objPath)) {
+            options.logger('Could not read image file at ' + imagePath + ' because it is outside of the obj directory and the secure flag is true. Material will ignore this image.');
+            return;
+        }
+        return loadImage(imagePath, options)
+            .then(function(image) {
+                if (defined(image)) {
+                    images[imagePath] = image;
+                }
+            })
+            .catch(function() {
+                options.logger('Could not read image file at ' + imagePath + '. Material will ignore this image.');
```

Comment:
OK, one comment about logging since it's on my mind.  When we do progress reporting, we should have well defined warnings (like type : MissingImage, path: 'expectPath.png' or something similar).  This way developers using the library can correctly detect and handle messages instead of relying on free-form strings.